<!doctype html>
<html lang="ja">
<head prefix="og: http://ogp.me/ns#">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
<title>edgeR: リードカウントから発現変動遺伝子を検出 - Heavy Watal</title>
<link rel="stylesheet" href="/css/theme.css">
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" href="/favicon-32x32.png" sizes="32x32">
<link rel="manifest" href="/manifest.json">
<meta name="theme-color" content="#e08010">
<meta name="author" content="Watal M. Iwasaki">
<meta property="og:title" content="edgeR: リードカウントから発現変動遺伝子を検出">
<meta property="og:type" content="article">
<meta property="og:url" content="https://heavywatal.github.io/rstats/edger.html">
<meta property="og:image" content="https://avatars.githubusercontent.com/heavywatal">
<meta property="og:description" content="">
<meta property="og:site_name" content="Heavy Watal">
<meta name="twitter:card" content="summary">
<meta name="twitter:site" content="@heavywatal">
<meta name="twitter:creator" content="@heavywatal">
<meta name="generator" content="Hugo 0.54-DEV" />
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.10.0/dist/katex.min.css" integrity="sha384-9eLZqc9ds8eNjO3TmqPeYcDj8n+Qfa4nuSiGYa6DjLNcv9BtN69ZIulL9+8CqC9Y" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.10.0/dist/katex.min.js" integrity="sha384-K3vbOmF2BtaVai+Qk37uypf7VrgBubhQreNQe9aGsz9lB63dIFiQVlJbr92dw2Lx" crossorigin="anonymous"></script>
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.10.0/dist/contrib/auto-render.min.js" integrity="sha384-kmZOZB5ObwgQnS/DuDg6TScgOiWWBiVt0plIRkZCmE6rDZGrEOQeHM5PcHi+nyqe" crossorigin="anonymous"
    onload="renderMathInElement(document.body);"></script>
<script>
document.addEventListener("DOMContentLoaded", function() {
  renderMathInElement(document.body, {
    delimiters: [
      {left: "\\[", right: "\\]", display: true},
      {left: "$", right: "$", display: false}
    ]
  });
});
</script>
<style>
.katex {
  font-size: 1.12em;
}

.katex-display > .katex {
  text-align: left;
  padding-left: 2rem;
}
</style>
<script type="application/ld+json">{
"@context": "http://schema.org",
"@type": "BreadcrumbList",
"itemListElement": [
 {"@type": "ListItem",
  "position": 1,
  "item": {
   "@id": "https://heavywatal.github.io/rstats.html",
   "name": "rstats"}},
 {"@type": "ListItem",
  "position": 2,
  "item": {
   "@id": "https://heavywatal.github.io/rstats/edger.html",
   "name": "edgeR"}}
]}</script>

<script defer src="https://use.fontawesome.com/releases/v5.6.1/js/brands.js"></script>
<script defer src="https://use.fontawesome.com/releases/v5.6.1/js/solid.js"></script>
<script defer src="https://use.fontawesome.com/releases/v5.6.1/js/fontawesome.js"></script>

<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
	(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
	m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
	})(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
	ga('create', 'UA-41178626-2', 'auto');
	
	ga('send', 'pageview');
}
</script>

</head>
<body>
<header><h1><a href="/">
<img class="logo" src="/heavywatal.svg" alt="岩嵜航">
Heavy Watal
</a></h1>
<form class="cse-search-box" action="/search.html">
<label><i class="fas fa-fw fa-search"></i><input type="text" name="q" aria-label="cse-search-box"></label>
</form>
</header>
<main>
<article>
<header><h1><a href="/rstats/edger.html">
edgeR — リードカウントから発現変動遺伝子を検出
</a></h1>
<nav class="tags"><ul>
<li><i class="fas fa-tag"></i> <a href="/tags/r.html">r</a></li>
<li><i class="fas fa-tag"></i> <a href="/tags/bioconductor.html">bioconductor</a></li>
</ul></nav>
</header>



<p><a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">https://bioconductor.org/packages/release/bioc/html/edgeR.html</a></p>

<p>Robinson MD, McCarthy DJ, Smyth GK (2010)
&ldquo;edgeR: a Bioconductor package for differential expression analysis of digital gene expression data.&rdquo;
<em>Bioinformatics</em> <strong>26</strong> (1):139&ndash;140
<a href="https://www.ncbi.nlm.nih.gov/pubmed/19910308">https://www.ncbi.nlm.nih.gov/pubmed/19910308</a></p>

<p><a href="/rstats/bioconductor.html">Bioconductor</a> パッケージとしてインストール</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r"><span class="kn">source</span><span class="p">(</span><span class="s">&#39;https://bioconductor.org/biocLite.R&#39;</span><span class="p">)</span>
biocLite<span class="p">(</span><span class="s">&#39;edgeR&#39;</span><span class="p">)</span></code></pre></div>
<p>ユーザーガイドPDFを開く</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r">edgeRUsersGuide<span class="p">()</span></code></pre></div>
<p>とても良くできたドキュメントなので必読</p>

<h2 id="使い方">使い方</h2>

<ol>
<li><code>HTSeq</code> などで遺伝子ごとのリードカウントを用意する</li>

<li><p>Rで読み込む</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r"><span class="kn">library</span><span class="p">(</span>edgeR<span class="p">)</span>

targets <span class="o">=</span> <span class="kt">data.frame</span><span class="p">(</span>group<span class="o">=</span><span class="kt">c</span><span class="p">(</span><span class="s">&#39;control&#39;</span><span class="p">,</span> <span class="s">&#39;case&#39;</span><span class="p">),</span>
                     files<span class="o">=</span><span class="kt">c</span><span class="p">(</span><span class="s">&#39;control.txt&#39;</span><span class="p">,</span> <span class="s">&#39;case.txt&#39;</span><span class="p">))</span>
dge <span class="o">=</span> readDGE<span class="p">(</span>targets<span class="p">,</span> header<span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span></code></pre></div></li>

<li><p>低カウント過ぎる遺伝子を除去</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r">ok_c <span class="o">=</span> <span class="p">(</span>dge<span class="o">$</span>counts <span class="o">&gt;</span> <span class="m">5</span><span class="p">)</span> <span class="o">%&gt;%</span> rowSums <span class="o">%&gt;%</span> <span class="p">{</span><span class="m">.</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">}</span>
ok_cpm <span class="o">=</span> dge <span class="o">%&gt;%</span> cpm <span class="o">%&gt;%</span> <span class="p">{</span><span class="m">.</span> <span class="o">&gt;</span> <span class="m">1</span><span class="p">}</span> <span class="o">%&gt;%</span> rowSums <span class="o">%&gt;%</span> <span class="p">{</span><span class="m">.</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">}</span>
dge <span class="o">=</span> dge<span class="p">[</span>ok_c <span class="o">&amp;</span> ok_cpm<span class="p">,</span> <span class="p">,</span> keep.lib.sizes<span class="o">=</span><span class="kc">FALSE</span><span class="p">]</span></code></pre></div></li>

<li><p>正規化係数を計算</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r">dge <span class="o">=</span> calcNormFactors<span class="p">(</span>dge<span class="p">)</span>
dge<span class="o">$</span>samples</code></pre></div></li>

<li><p>モデル</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r">design <span class="o">=</span> model.matrix<span class="p">(</span><span class="o">~</span> <span class="m">0</span> <span class="o">+</span> group<span class="p">,</span> data<span class="o">=</span>targets<span class="p">)</span></code></pre></div></li>

<li><p>common dispersion, trended dispersion, tagwise dispersionを推定</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r">dge <span class="o">=</span> estimateDisp<span class="p">(</span>dge<span class="p">,</span> design<span class="p">)</span></code></pre></div>
<p>ただしこれにはbiological replicatesが必要。
1グループ1サンプルずつしか無い場合は4つの選択肢がある。</p>

<ol>
<li>検定を諦めてdescriptive discussion (推奨)</li>

<li><p>経験的な値を適当に入れとく</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r">bcv <span class="o">=</span> <span class="m">0.4</span>  <span class="c1"># for human</span>
bcv <span class="o">=</span> <span class="m">0.1</span>  <span class="c1"># for genetically identical model organisms</span>
bcv <span class="o">=</span> <span class="m">0.01</span> <span class="c1"># for technical replicates</span>
dge<span class="o">$</span>common.dispersion <span class="o">=</span> bcv <span class="o">^</span> <span class="m">2</span></code></pre></div></li>

<li><p>説明変数を減らしたモデルでdispersionを推定し、フルモデルで使う</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r">reduced <span class="o">=</span> estimateGLMCommonDisp<span class="p">(</span>dge<span class="p">,</span> reduced.design<span class="p">,</span> method<span class="o">=</span><span class="s">&#39;deviance&#39;</span><span class="p">,</span> robust<span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> subset<span class="o">=</span><span class="kc">NULL</span><span class="p">)</span>
dge<span class="o">$</span>common.dispersion <span class="o">=</span> reduced<span class="o">$</span>common.dispersion</code></pre></div>
<p>合計カウントが多くてDEGが比較的少ないときに有効。
当然、1変数2群比較では使えない。</p></li>

<li><p>群間で差がないと考えられるhousekeeping genesから推定。
100個以上の遺伝子を使うのが望ましい。
例えばヒトなら</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r"><span class="kp">system</span><span class="p">(</span><span class="s">&#39;wget http://www.tau.ac.il/~elieis/HKG/HK_genes.txt&#39;</span><span class="p">)</span>
hk_genes <span class="o">=</span> read_tsv<span class="p">(</span><span class="s">&#39;HK_genes.txt&#39;</span><span class="p">,</span> col_names<span class="o">=</span><span class="kt">c</span><span class="p">(</span><span class="s">&#39;gene_symbol&#39;</span><span class="p">,</span> <span class="s">&#39;refseq&#39;</span><span class="p">))</span>
tmp <span class="o">=</span> dge
tmp<span class="o">$</span>samples<span class="o">$</span>group <span class="o">=</span> <span class="m">1</span>
housekeeping <span class="o">=</span> <span class="kp">rownames</span><span class="p">(</span>tmp<span class="o">$</span>counts<span class="p">)</span> <span class="o">%in%</span> hk_genes<span class="o">$</span>gene_symbol
hk <span class="o">=</span> estimateCommonDisp<span class="p">(</span>tmp<span class="p">[</span>housekeeping<span class="p">,])</span>
dge<span class="o">$</span>common.dispersion <span class="o">=</span> hk<span class="o">$</span>common.dispersion</code></pre></div></li>
</ol></li>

<li><p>1変数2群ならシンプルに検定</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r">et <span class="o">=</span> exactTest<span class="p">(</span>dge<span class="p">)</span>
topTags<span class="p">(</span>et<span class="p">,</span> <span class="m">20</span><span class="p">,</span> adjust.method<span class="o">=</span><span class="s">&#39;BH&#39;</span><span class="p">,</span> sort.by<span class="o">=</span><span class="s">&#39;PValue&#39;</span><span class="p">)</span></code></pre></div>
<p>多群ならGLMで尤度比検定</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r">fit <span class="o">=</span> glmFit<span class="p">(</span>dge<span class="p">,</span> design<span class="p">)</span>
lrt <span class="o">=</span> glmLRT<span class="p">(</span>fit<span class="p">,</span> coef<span class="o">=</span><span class="m">2</span><span class="o">:</span><span class="m">3</span><span class="p">)</span>

lrt_1vs2 <span class="o">=</span> glmLRT<span class="p">(</span>fit<span class="p">,</span> coef<span class="o">=</span><span class="m">2</span><span class="p">)</span>
lrt_1vs3 <span class="o">=</span> glmLRT<span class="p">(</span>fit<span class="p">,</span> coef<span class="o">=</span><span class="m">3</span><span class="p">)</span>
lrt_2vs3 <span class="o">=</span> glmLRT<span class="p">(</span>fit<span class="p">,</span> contrast<span class="o">=</span><span class="kt">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span> <span class="m">-1</span><span class="p">,</span> <span class="m">1</span><span class="p">))</span></code></pre></div>
<p>検定結果からDEGを抜き出す</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r">min_lfc <span class="o">=</span> <span class="m">1</span>
de <span class="o">=</span> decideTestsDGE<span class="p">(</span>et<span class="p">,</span> adjust.method<span class="o">=</span><span class="s">&#39;BH&#39;</span><span class="p">,</span> p.value<span class="o">=</span><span class="m">0.01</span><span class="p">,</span> lfc<span class="o">=</span>min_lfc<span class="p">)</span>
de_tags <span class="o">=</span> <span class="kp">rownames</span><span class="p">(</span>dge<span class="p">)[</span><span class="kp">as.logical</span><span class="p">(</span>de<span class="p">)]</span></code></pre></div>
<p>P値とlogFCの両方で切ったほうがいいらしい。</p>

<div class="note">
<p><code>topTags()</code> や <code>decideTestsDGE()</code>
で多重検定の補正をするということは、
<code>et</code> や <code>lrt</code> の <code>PValue</code> は補正前。</p>

</div>
</li>

<li><p>プロット <code>logFC ~ mean(logCPM)</code> してみる</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r">plotSmear<span class="p">(</span>et<span class="p">,</span> de.tags<span class="o">=</span>de_tags<span class="p">)</span>
abline<span class="p">(</span>h<span class="o">=</span>min_lfc <span class="o">*</span> <span class="kt">c</span><span class="p">(</span><span class="m">-1</span><span class="p">,</span> <span class="m">1</span><span class="p">),</span> col<span class="o">=</span><span class="s">&#34;blue&#34;</span><span class="p">)</span></code></pre></div></li>

<li><p>Gene Ontology解析</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r">go <span class="o">=</span> goana<span class="p">(</span>lrt<span class="p">,</span> species<span class="o">=</span><span class="s">&#39;Hs&#39;</span><span class="p">)</span>
topGO<span class="p">(</span>go<span class="p">)</span></code></pre></div>
<p>NCBI RefSeqアノテーションを使うのでIDはEntrez系で</p></li>
</ol>

<h2 id="appendix">Appendix</h2>

<p>レプリケート <em>i</em> における遺伝子 <em>g</em> の観察リード数を $y _{gi}$、
知りたい真の発現fractionを $\pi _{gi}$ とする。</p>

<div>\[\begin{aligned}
\sum _g \pi _{gi} &= 1 \\
\sqrt {\phi _g} &\equiv \text{CV}[\pi _{gi}]_i
                = \frac {\text{sd}[\pi _{gi}]_i} {\text{mean}[\pi _{gi}]_i} \\
N_i \pi _{gi} &\sim \text{Gamma}(\phi _g^{-1}, \mu _{gi} \phi _g) \\
\mathrm E[N_i \pi _{gi}] &= k\theta = \mu _{gi} \\
\text{var}[N_i \pi _{gi}] &= k\theta^2 = \mu _{gi}^2 \phi _g \\
y_{gi} &\sim \text{Poisson}(N_i \pi _{gi}) \\
       &= \int _0^\infty \text{Poisson}(N_i \pi _{gi})~
                         \text{Gamma}(\phi _g^{-1}, \mu _{gi} \phi _g)~ \mathrm d N_i \pi _{gi} \\
       &= \text{NB}(\phi _g^{-1}, \frac {\mu _{gi} \phi _g} {1 + \mu _{gi} \phi _g}) \\
\mathrm E[y_{gi}] &= \mu _{gi} \\
\text{var}[y_{gi}] &= \mathrm E \big[\text{var}[y_{gi} \mid \pi _{gi}]_i \big]_\pi
                      + \text{var} \big[\mathrm E[y_{gi}\mid \pi _{gi}]_i \big]_\pi \\
                     &= \mathrm E[N _i \pi _{gi}] _\pi + \text{var}[N _i \pi _{gi}] _\pi \\
                     &= \mu _{gi} + \mu _{gi}^2 \phi _g \\
\text{CV}^2[y_{gi}] &= 1 / \mu _{gi} + \phi _g \\
                   &= \text{CV}^2[y_{gi} \mid \pi _{gi}] + \text{CV}^2[\pi _{gi}] \\
                   &= \text{Technical~CV}^2 + \text{Biological~CV}^2 \\
\end{aligned}\]</div>

<dl>
<dt>dispersion $\phi _g$</dt>
<dd>普通は $D = \sigma^2 / \mu$ と定義されるけど、
ここでは $\text{CV}^2$</dd>
<dt>BCV $\sqrt{\phi _g}$</dt>
<dd>biological coefficient of variation.
こっちは普通と同じように $\text{CV} = \sigma / \mu$ 。
sequencing depthがどんなに大きくても残るばらつき。</dd>
<dt>DGE</dt>
<dd>digital gene expression</dd>
<dt>CPM</dt>
<dd>counts per million</dd>
<dt>logFC</dt>
<dd>log2-fold-change</dd>
</dl>

<hr>

<p><a href="https://www.amazon.co.jp/dp/4621062506//ref=as_li_ss_il?ie=UTF8&linkCode=li3&tag=heavywatal-22&linkId=c58e6e9dc365558cc336d9ea0a2c8a12" target="_blank"><img border="0" src="//ws-fe.amazon-adsystem.com/widgets/q?_encoding=UTF8&ASIN=4621062506&Format=_SL250_&ID=AsinImage&MarketPlace=JP&ServiceVersion=20070822&WS=1&tag=heavywatal-22" ></a><img src="https://ir-jp.amazon-adsystem.com/e/ir?t=heavywatal-22&l=li3&o=9&a=4621062506" width="1" height="1" border="0" alt="" style="border:none !important; margin:0px !important;" />
<a href="https://www.amazon.co.jp/dp/4320057082//ref=as_li_ss_il?ie=UTF8&linkCode=li3&tag=heavywatal-22&linkId=b0a0cf1dfe769f34f7544db70a0f6711" target="_blank"><img border="0" src="//ws-fe.amazon-adsystem.com/widgets/q?_encoding=UTF8&ASIN=4320057082&Format=_SL250_&ID=AsinImage&MarketPlace=JP&ServiceVersion=20070822&WS=1&tag=heavywatal-22" ></a><img src="https://ir-jp.amazon-adsystem.com/e/ir?t=heavywatal-22&l=li3&o=9&a=4320057082" width="1" height="1" border="0" alt="" style="border:none !important; margin:0px !important;" />
<a href="https://www.amazon.co.jp/dp/4320123700//ref=as_li_ss_il?ie=UTF8&linkCode=li3&tag=heavywatal-22&linkId=f7d9a9e5ba94b0fa7da4a582b70da85a" target="_blank"><img border="0" src="//ws-fe.amazon-adsystem.com/widgets/q?_encoding=UTF8&ASIN=4320123700&Format=_SL250_&ID=AsinImage&MarketPlace=JP&ServiceVersion=20070822&WS=1&tag=heavywatal-22" ></a><img src="https://ir-jp.amazon-adsystem.com/e/ir?t=heavywatal-22&l=li3&o=9&a=4320123700" width="1" height="1" border="0" alt="" style="border:none !important; margin:0px !important;" /></p>

</article>
</main>

<nav class="menu">

<div><a href="/about.html">About</a></div>

<div><a href="/research.html">Research Interests</a></div>

<input type="checkbox" id="menu-rstats" checked>
<label for="menu-rstats" class="active">R stats</label>
<ul>
<li><a href="/rstats/intro.html">R自学自習の基礎知識</a></li>
<li><a href="/rstats/programming.html">RプログラミングTips</a></li>
<li><a href="/rstats/config.html">R環境設定</a></li>
<li><a href="/rstats/ggplot2.html">ggplot2</a></li>
<li><a href="/rstats/dplyr.html">dplyr</a></li>
<li><a href="/rstats/purrr.html">purrr</a></li>
<li><a href="/rstats/tidyr.html">tidyr</a></li>
<li><a href="/rstats/readr.html">readr</a></li>
<li><a href="/rstats/stringr.html">stringr</a></li>
<li><a href="/rstats/foreach.html">foreach/parallel</a></li>
<li><a href="/rstats/devtools.html">devtools</a></li>
<li><a href="/rstats/rcpp.html">Rcpp</a></li>
<li><a href="/rstats/bioconductor.html">Bioconductor</a></li>
<li><a href="/rstats/rtracklayer.html">rtracklayer</a></li>
<li><a href="/rstats/biomart.html">biomaRt</a></li>
<li><a href="/rstats/ggbio.html">ggbio</a></li>
<li class="active"><a href="/rstats/edger.html">edgeR</a></li>
<li><a href="/rstats/topgo.html">topGO</a></li>
<li><a href="/rstats/stan.html">Stan</a></li>
<li><a href="/rstats/rgl.html">rgl</a></li>
<li><a href="/rstats/plyr.html">plyr</a></li>
<li><a href="/rstats/reshape2.html">reshape2</a></li>
</ul>

<input type="checkbox" id="menu-python">
<label for="menu-python">Python</label>
<ul>
<li><a href="/python/install.html">Pythonインストール</a></li>
<li><a href="/python/pip.html">pip</a></li>
<li><a href="/python/ipython.html">IPython</a></li>
<li><a href="/python/biopython.html">BioPython</a></li>
<li><a href="/python/scipy.html">NumPy, SciPy</a></li>
<li><a href="/python/pandas.html">Pandas</a></li>
<li><a href="/python/pyqt.html">PyQt</a></li>
<li><a href="/python/concurrent.html">concurrent.futures</a></li>
<li><a href="/python/copy.html">copy</a></li>
<li><a href="/python/matplotlib.html">matplotlib &#43; seaborn</a></li>
<li><a href="/python/setuptools.html">setuptools</a></li>
</ul>

<input type="checkbox" id="menu-cxx">
<label for="menu-cxx">C&#43;&#43;</label>
<ul>
<li><a href="/cxx/boost.html">Boost</a></li>
<li><a href="/cxx/getopt.html">C&#43;&#43;コマンドライン引数</a></li>
<li><a href="/cxx/speed.html">C&#43;&#43;高速化</a></li>
<li><a href="/cxx/clang.html">clang / llvm</a></li>
<li><a href="/cxx/gcc.html">gcc</a></li>
<li><a href="/cxx/bitwise.html">ビット演算</a></li>
<li><a href="/cxx/random.html">擬似乱数生成器</a></li>
</ul>

<input type="checkbox" id="menu-bio">
<label for="menu-bio">Biology</label>
<ul>
<li><a href="/bio/blast.html">BLAST</a></li>
<li><a href="/bio/gene_ontology.html">Gene Ontology</a></li>
<li><a href="/bio/meme.html">MEME</a></li>
<li><a href="/bio/popgen.html">Population Genetics</a></li>
<li><a href="/bio/repeatmasker.html">RepeatMasker</a></li>
<li><a href="/bio/samtools.html">SAMtools</a></li>
<li><a href="/bio/shirokane.html">SHIROKANE</a></li>
<li><a href="/bio/stochastic_process.html">Stochastic Process</a></li>
<li><a href="/bio/dadi.html">dadi</a></li>
<li><a href="/bio/mathmorph.html">数理形態学</a></li>
<li><a href="/bio/linear_algebra.html">線形代数</a></li>
<li><a href="/bio/complexnetwork.html">複雑ネットワーク</a></li>
<li><a href="/bio/nig.html">遺伝研スパコン</a></li>
<li><a href="/bio/motif.html">配列モチーフ探索</a></li>
</ul>

<input type="checkbox" id="menu-dev">
<label for="menu-dev">Developer Tools</label>
<ul>
<li><a href="/dev/atom.html">Atom</a></li>
<li><a href="/dev/cmake.html">CMake</a></li>
<li><a href="/dev/emacs.html">Emacs</a></li>
<li><a href="/dev/git.html">Git</a></li>
<li><a href="/dev/hugo.html">Hugo</a></li>
<li><a href="/dev/mercurial.html">Mercurial</a></li>
<li><a href="/dev/qt.html">Qt</a></li>
<li><a href="/dev/torque.html">TORQUE</a></li>
<li><a href="/dev/autotools.html">autoconf, automake</a></li>
<li><a href="/dev/make.html">make</a></li>
<li><a href="/dev/mount.html">mount</a></li>
<li><a href="/dev/nano.html">nano</a></li>
<li><a href="/dev/rsync.html">rsync</a></li>
<li><a href="/dev/ssh.html">ssh</a></li>
<li><a href="/dev/tmux.html">tmux</a></li>
<li><a href="/dev/vi.html">vi</a></li>
<li><a href="/dev/zsh.html">zsh</a></li>
<li><a href="/dev/sh.html">シェルスクリプト</a></li>
<li><a href="/dev/nohup.html">プロセス管理</a></li>
<li><a href="/dev/devenv.html">開発環境</a></li>
</ul>

<input type="checkbox" id="menu-mac">
<label for="menu-mac">Mac</label>
<ul>
<li><a href="/mac/applescript.html">AppleScript</a></li>
<li><a href="/mac/homebrew.html">Homebrew</a></li>
<li><a href="/mac/keyboard.html">Keyboard</a></li>
<li><a href="/mac/macports.html">MacPorts</a></li>
<li><a href="/mac/command.html">Mac固有コマンド</a></li>
<li><a href="/mac/quicklook.html">QuickLook</a></li>
<li><a href="/mac/spotlight.html">Spotlight</a></li>
<li><a href="/mac/winebottler.html">WineBottler</a></li>
<li><a href="/mac/kotoeri.html">ことえり</a></li>
</ul>

<input type="checkbox" id="menu-lectures">
<label for="menu-lectures">Lectures</label>
<ul>
<li><a href="/lectures/verhoef_comm_ecol-11.html">Community Ecology 輪読会 11章</a></li>
<li><a href="/lectures/git2018nrifs.html">Git入門2018</a></li>
<li><a href="/lectures/prml-11-1.html">PRML輪読会 11章1節</a></li>
<li><a href="/lectures/prml-2-1.html">PRML輪読会 2章前半</a></li>
<li><a href="/lectures/prml-3-4.html">PRML輪読会 3章4節</a></li>
<li><a href="/lectures/wakeley-2-2.html">Wakeley輪読会 2章2節</a></li>
</ul>

<input type="checkbox" id="menu-misc">
<label for="menu-misc">Miscellaneous</label>
<ul>
<li><a href="/misc/fonts.html">Fonts</a></li>
<li><a href="/misc/latex.html">LaTeX</a></li>
</ul>

<div><a href="/tags.html">Tags</a></div>
</nav>
<footer><small>(ɔ) 2008 岩嵜航 Watal M. Iwasaki</small></footer>
</body>
</html>
