<!doctype html>
<html>
<head prefix="og: http://ogp.me/ns#">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
<title>topGO: Bioconductor でenrichment解析 - Heavy Watal</title>
<link rel="stylesheet" href="/css/theme.css">
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" href="/favicon-32x32.png" sizes="32x32">
<link rel="manifest" href="/manifest.json">
<meta name="theme-color" content="#e08010">
<meta name="author" content="Watal M. Iwasaki">
<meta property="og:title" content="topGO: Bioconductor でenrichment解析">
<meta property="og:type" content="article">
<meta property="og:url" content="https://heavywatal.github.io/rstats/topgo.html">
<meta property="og:image" content="https://avatars.githubusercontent.com/heavywatal">
<meta property="og:description" content="">
<meta property="og:site_name" content="Heavy Watal">
<meta name="twitter:card" content="summary">
<meta name="twitter:site" content="@heavywatal">
<meta name="twitter:creator" content="@heavywatal">
<meta name="generator" content="Hugo 0.92.2" />
<link rel="stylesheet" href="/lib/katex/katex.min.css">
<script defer src="/lib/katex/katex.min.js"></script>
<script defer src="/lib/katex/contrib/auto-render.min.js"></script>
<script>
document.addEventListener("DOMContentLoaded", function() {
  renderMathInElement(document.body, {
    delimiters: [
      {left: "\\[", right: "\\]", display: true},
      {left: "$", right: "$", display: false}
    ]
  });
});
</script>
<style>
.katex {
  font-size: 1.12em;
}

.katex-display > .katex {
  text-align: left;
  padding-left: 2rem;
}
</style>
<script type="application/ld+json">{
"@context": "http://schema.org",
"@type": "BreadcrumbList",
"itemListElement": [
 {"@type": "ListItem",
  "position": 1,
  "item": {"@id":"https://heavywatal.github.io/rstats.html","name":"rstats"} },
 {"@type": "ListItem",
  "position": 2,
  "item": {"@id":"https://heavywatal.github.io/rstats/topgo.html","name":"topGO"} }
]}</script>

<script defer src="https://use.fontawesome.com/releases/v5.8.2/js/all.js" integrity="sha384-DJ25uNYET2XCl5ZF++U8eNxPWqcKohUUBUpKGlNLMchM7q4Wjg2CUpjHLaL8yYPH" crossorigin="anonymous"></script>

<script async src="https://www.googletagmanager.com/gtag/js?id=G-V60H2JH0G6"></script>
<script>
var doNotTrack = false;
if (!doNotTrack) {
	window.dataLayer = window.dataLayer || [];
	function gtag(){dataLayer.push(arguments);}
	gtag('js', new Date());
	gtag('config', 'G-V60H2JH0G6', { 'anonymize_ip': false });
}
</script>

</head>
<body>
<header><h1><a href="/">
<img class="logo" src="/heavywatal.svg" alt="岩嵜航">
Heavy Watal
</a></h1>
</header>
<main>
<article>
<header><h1><a href="/rstats/topgo.html">
topGO — Bioconductor でenrichment解析
</a></h1>
<nav class="tags"><ul>
<li><i class="fas fa-tag"></i> <a href="/tags/r.html">r</a></li>
<li><i class="fas fa-tag"></i> <a href="/tags/bioconductor.html">bioconductor</a></li>
</ul></nav>
</header>

<p><a href="/bio/gene_ontology.html">Gene Ontology</a> を利用して、
ある遺伝子セットにどんな機能が多めに含まれているかを解析する。</p>
<p><a href="https://www.bioconductor.org/packages/release/bioc/html/topGO.html">https://www.bioconductor.org/packages/release/bioc/html/topGO.html</a></p>
<p>Rの中からインストール</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&#34;topGO&#34;</span><span class="p">)</span>
</code></pre></div><h2 id="使い方">使い方</h2>
<p>ヒト遺伝子にランダムなスコアをつけた架空データ</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">)</span>
<span class="n">entrez_ids</span> <span class="o">=</span> <span class="nf">mappedkeys</span><span class="p">(</span><span class="n">org.Hs.egGO</span><span class="p">)</span>
<span class="n">scores</span> <span class="o">=</span> <span class="nf">runif</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">entrez_ids</span><span class="p">),</span> <span class="m">0</span><span class="p">,</span> <span class="m">1</span><span class="p">)</span>  <span class="c1"># p-value-like</span>
<span class="c1"># scores = rnorm(length(entrez_ids), 0, 0.4)  # log2FC-like</span>
<span class="n">named_scores</span> <span class="o">=</span> <span class="nf">setNames</span><span class="p">(</span><span class="n">scores</span><span class="p">,</span> <span class="n">entrez_ids</span><span class="p">)</span>
</code></pre></div><h3 id="topgodata-を作る"><code>topGOdata</code> を作る</h3>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="n">tg_data</span> <span class="o">=</span> <span class="nf">new</span><span class="p">(</span><span class="s">&#34;topGOdata&#34;</span><span class="p">,</span>
    <span class="n">ontology</span><span class="o">=</span><span class="s">&#34;BP&#34;</span><span class="p">,</span>
    <span class="n">allGenes</span><span class="o">=</span><span class="n">named_scores</span><span class="p">,</span>
    <span class="n">geneSelectionFun</span><span class="o">=</span><span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span><span class="n">x</span> <span class="o">&lt;</span> <span class="m">0.01</span><span class="p">},</span>
    <span class="c1"># geneSelectionFun=function(x) {abs(x) &gt; 1},</span>
    <span class="n">nodeSize</span><span class="o">=</span><span class="m">10</span><span class="p">,</span>
    <span class="n">annotationFun</span><span class="o">=</span><span class="n">annFUN.org</span><span class="p">,</span>
    <span class="n">mapping</span><span class="o">=</span><span class="s">&#34;org.Hs.eg.db&#34;</span><span class="p">,</span>
    <span class="n">ID</span><span class="o">=</span><span class="s">&#34;entrez&#34;</span><span class="p">)</span>
</code></pre></div><dl>
<dt><code>ontology</code></dt>
<dd><code>BP</code>, <code>CC</code>, <code>MF</code> のどれか</dd>
<dt><code>description</code> (省略可)</dt>
<dd>説明string</dd>
<dt><code>allGenes</code></dt>
<dd>遺伝子名を名前とするnamed vector。
値はp-valueとかlog2FCとかなんとか、好きなスコア。</dd>
<dd>名前は&quot;all&quot;だけど常に全遺伝子である必要はなくて、
むしろ目的に応じて適切な&quot;background&quot;を選んで渡すべき。</dd>
<dt><code>geneSelectionFun</code></dt>
<dd><code>allGenes</code> の数値を引数として今回興味のある遺伝子に <code>TRUE</code> を返すような関数。
スコアが閾値より大きい・小さいとか、上位100までとか。
<code>function(p) {p &lt; 0.01}</code>,
<code>function(log2FC) {abs(log2FC) &gt; 1}</code>,
<code>function(score) {rank(score) &lt;= 100L}</code></dd>
<dd>必要なのは <code>statistic=&quot;fisher&quot;</code> のときだけで、なおかつ実際に適用されるのは
<code>runTest()</code> 実行時なのにここで入力させる、という筋の悪いデザイン。
スコアをそのまま使うはずの <code>statistic=&quot;ks&quot;</code> などでも要求され、
Significant genes という謎の結果が計算されてしまう。
fisherを使わない場合は結果の誤解を防ぐためにも一律 <code>FALSE</code> を返す関数にしておくのが安全。
<code>function(x) {logical(length(x))}</code></dd>
<dt><code>nodeSize</code></dt>
<dd>紐付けられた遺伝子の数がこれより少ないGO termを除外。1以上の整数。</dd>
<dt><a href="https://www.rdocumentation.org/packages/topGO/topics/annFUN"><code>annotationFun</code></a></dt>
<dd>遺伝子IDとGO termを結びつける関数。</dd>
<dd>後述のように <code>org.**.**.db</code> パッケージがあるようなメジャー種の遺伝子IDなら <code>annFUN.org</code> 。
チップが登録されているマイクロアレイなら <code>annFUN.db</code> 。
自作マップで頑張るなら <code>annFUN.gene2GO</code>, <code>annFUN.GO2genes</code>, <code>annFUN.file</code> 。</dd>
<dt><code>...</code></dt>
<dd>以降は <code>annotationFun()</code> に渡す引数。</dd>
<dd>e.g., <code>annFUN.org(whichOnto, feasibleGenes, mapping, ID=&quot;entrez&quot;)</code>
<dl>
<dt><code>mapping</code></dt>
<dd>IDマッピング用のBioConductorパッケージ。
<a href="https://bioconductor.org/packages/release/data/annotation/">BioConductor AnnotationData Packages</a>
から探す。例えばヒトなら <code>org.Hs.eg.db</code> 。</dd>
<dt><code>ID</code></dt>
<dd><code>allGenes</code> に与えた名前の種類。</dd>
<dd><code>entrez</code>, <code>genbank</code>, <code>alias</code>, <code>ensembl</code>,
<code>symbol</code>, <code>genename</code>, <code>unigene</code>.
(<code>annFUN.org()</code> の中に書いてある。case-insensitive)</dd>
</dl>
</dd>
</dl>
<h3 id="検定">検定</h3>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="nf">whichAlgorithms</span><span class="p">()</span>
<span class="nf">whichTests</span><span class="p">()</span>
<span class="n">so</span><span class="o">=</span><span class="s">&#34;increasing&#34;</span>
<span class="n">resClassicFisher</span> <span class="o">=</span> <span class="nf">runTest</span><span class="p">(</span><span class="n">tg_data</span><span class="p">,</span> <span class="n">algorithm</span><span class="o">=</span><span class="s">&#34;classic&#34;</span><span class="p">,</span> <span class="n">statistic</span><span class="o">=</span><span class="s">&#34;fisher&#34;</span><span class="p">,</span> <span class="n">sortOrder</span><span class="o">=</span><span class="n">so</span><span class="p">)</span>
<span class="n">resElimFisher</span> <span class="o">=</span> <span class="nf">runTest</span><span class="p">(</span><span class="n">tg_data</span><span class="p">,</span> <span class="n">algorithm</span><span class="o">=</span><span class="s">&#34;elim&#34;</span><span class="p">,</span> <span class="n">statistic</span><span class="o">=</span><span class="s">&#34;fisher&#34;</span><span class="p">,</span> <span class="n">sortOrder</span><span class="o">=</span><span class="n">so</span><span class="p">)</span>
<span class="n">resWeightFisher</span> <span class="o">=</span> <span class="nf">runTest</span><span class="p">(</span><span class="n">tg_data</span><span class="p">,</span> <span class="n">algorithm</span><span class="o">=</span><span class="s">&#34;weight&#34;</span><span class="p">,</span> <span class="n">statistic</span><span class="o">=</span><span class="s">&#34;fisher&#34;</span><span class="p">,</span> <span class="n">sortOrder</span><span class="o">=</span><span class="n">so</span><span class="p">)</span>
<span class="n">resClassicKS</span> <span class="o">=</span> <span class="nf">runTest</span><span class="p">(</span><span class="n">tg_data</span><span class="p">,</span> <span class="n">algorithm</span><span class="o">=</span><span class="s">&#34;classic&#34;</span><span class="p">,</span> <span class="n">statistic</span><span class="o">=</span><span class="s">&#34;ks&#34;</span><span class="p">,</span> <span class="n">sortOrder</span><span class="o">=</span><span class="n">so</span><span class="p">)</span>
<span class="n">resElimKS</span> <span class="o">=</span> <span class="nf">runTest</span><span class="p">(</span><span class="n">tg_data</span><span class="p">,</span> <span class="n">algorithm</span><span class="o">=</span><span class="s">&#34;elim&#34;</span><span class="p">,</span> <span class="n">statistic</span><span class="o">=</span><span class="s">&#34;ks&#34;</span><span class="p">,</span> <span class="n">sortOrder</span><span class="o">=</span><span class="n">so</span><span class="p">)</span>
<span class="n">resWeightKS</span> <span class="o">=</span> <span class="nf">runTest</span><span class="p">(</span><span class="n">tg_data</span><span class="p">,</span> <span class="n">algorithm</span><span class="o">=</span><span class="s">&#34;weight01&#34;</span><span class="p">,</span> <span class="n">statistic</span><span class="o">=</span><span class="s">&#34;ks&#34;</span><span class="p">,</span> <span class="n">sortOrder</span><span class="o">=</span><span class="n">so</span><span class="p">)</span>
</code></pre></div><dl>
<dt><code>algorithm</code></dt>
<dd><code>classic</code>: GO termをそのまま使って計算。偽陽性多め。</dd>
<dd><code>elim</code> <a href="https://doi.org/10.1093/bioinformatics/btl140">(Alexa et al. 2006)</a>:
DAG上での隣接関係を考慮して補正。
下位termから検定を始め、有意なものが見つかったらそこに含まれる遺伝子を祖先ノードから除外する。
classicに比べて上位termの偽陽性が減ってconservative。
<code>cutOff</code> オプションでこの判定基準を 0.01 から変更可能。</dd>
<dd><code>weight</code> (Alexa et al. 2006):
elimを一般化して少しマイルドにしたような感じ。
有意な子ノードを多く持つ親ノードは生き残る。
classicよりconservativeだがelimより取りこぼさない。
計算が複雑すぎるせいか検定は <code>fisher</code> しかサポートされていない。</dd>
<dd><code>weight01</code>: &ldquo;mixture between the <code>elim</code> and the <code>weight</code> algorithms&rdquo;
とのことだが詳細は不明。
topGOデフォルトに据えるくらい自信あるんだろうけど。検定は全て可能。</dd>
<dd><code>lea</code>: ドキュメントでも論文でも言及無し。お蔵入りしたプロトタイプ？</dd>
<dd><code>parentChild</code> <a href="https://doi.org/10.1093/bioinformatics/btm440">(Grossmann et al. 2007)</a>:
classicよりも下位termでの偽陽性が少ない。
親を複数持つ場合の扱い2つ(union or intercection)のうちどちらを採用してるかは不明。
検定は <code>fisher</code> のみ。</dd>
<dt><code>statistic</code></dt>
<dd><code>fisher</code>: 遺伝子の数に基づいて検定。セットに含まれているか否かの二値。</dd>
<dd><code>ks</code> <a href="https://doi.org/10.1186/1471-2105-10-47">(Ackermann and Strimmer 2009)</a>:
遺伝子のスコアや順位に基づいて検定。
閾値で遺伝子セットを区切らずに済む。
(普通のKSだったら上位への偏りだけを見ているとは限らないけどそのあたりは調整済み？)</dd>
<dd><code>globaltest</code> <a href="https://doi.org/10.1093/bioinformatics/btm051">(Goeman and Bühlmann 2007)</a>:
統計量を挟まず生データに基づいて検定。
(<code>topGOdata</code> にどうやってデータ渡すんだろう？)</dd>
<dd><code>t</code>, <code>sum</code>, <code>ks.ties</code>: 不明。</dd>
<dt><code>scoreOrder</code></dt>
<dd><code>allGenes</code> に与えた値が小さいほど良いP値とかなら <code>increasing</code> (デフォルト)。
興味のある遺伝子で値が高くなるlog2FCのようなスコアなら <code>decreasing</code> を指定。
ドキュメントにあんまりちゃんと載ってない？</dd>
<dt>結局どれを使うか？</dt>
<dd><code>algorithm</code> はとりあえず最もconservativeで検定も自由な <code>elim</code> 。
もしくは作者Alexaさんを信じて <code>weight01</code> 。デフォルト設定は論文にも書きやすい。</dd>
<dd><code>statistic</code> は、遺伝子セットが既に区切ってあるなら <code>fisher</code> 、
DEG解析やら何やらで遺伝子がp値やスコアを持ってるなら <code>ks</code> 。
<code>globaltest</code> を使いたい場合は
<a href="http://bioconductor.org/packages/release/bioc/html/globaltest.html">本家 <code>globaltest</code> パッケージ</a>
を参照。</dd>
</dl>
<p>結果は <code>topGOresult</code> オブジェクト。
計算されるp値は多重検定の補正をしていない生の値。
&ldquo;Significant genes&rdquo; は <code>geneSelectionFun(allGenes)</code> で <code>TRUE</code> になったものの数で、
無関係なはずのKSでも計算されてしまって気持ち悪い。</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="nf">str</span><span class="p">(</span><span class="n">resElimKS</span><span class="p">)</span>
<span class="nf">geneData</span><span class="p">(</span><span class="n">resElimKS</span><span class="p">)</span>
<span class="nf">score</span><span class="p">(</span><span class="n">resElimKS</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">head</span><span class="p">()</span>
</code></pre></div><h3 id="解釈描画">解釈・描画</h3>
<p><code>GenTable()</code> を使って GO term Over-representation ランキングを表示できる。
複数の <code>runTest()</code> 産物を好きな名前で複数並べたりすることもできる。
が、P値が桁の小さい文字列型になってたり勝手に行が削られたりして怖いので使わない。
<code>GenTable()</code> の実装を参考に似た形式のより良いテーブルを自分で作る。</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="c1"># num_nodes=length(tg_data@graph@nodes)  # GO terms &gt;nodeSize</span>
<span class="c1"># tg_table = GenTable(tg_data,</span>
<span class="c1">#   classicFisher=resClassicFisher,</span>
<span class="c1">#   elimFisher=resElimFisher,</span>
<span class="c1">#   classicKS=resClassicKS,</span>
<span class="c1">#   elimKS=resElimKS,</span>
<span class="c1">#   topNodes=num_nodes) %&gt;%</span>
<span class="c1">#   tibble::as_tibble() %&gt;%</span>
<span class="c1">#   print()</span>

<span class="n">annoStat</span> <span class="o">=</span> <span class="nf">termStat</span><span class="p">(</span><span class="n">tg_data</span><span class="p">,</span> <span class="nf">sort</span><span class="p">(</span><span class="n">tg_data</span><span class="o">@</span><span class="n">graph</span><span class="o">@</span><span class="n">nodes</span><span class="p">))</span> <span class="o">%&gt;%</span>
  <span class="n">tibble</span><span class="o">::</span><span class="nf">rownames_to_column</span><span class="p">(</span><span class="n">var</span> <span class="o">=</span> <span class="s">&#34;GO.ID&#34;</span><span class="p">)</span> <span class="o">%&gt;%</span>
  <span class="n">tibble</span><span class="o">::</span><span class="nf">as_tibble</span><span class="p">()</span> <span class="o">%&gt;%</span>
  <span class="n">dplyr</span><span class="o">::</span><span class="nf">mutate</span><span class="p">(</span><span class="n">Term</span><span class="o">=</span><span class="n">topGO</span><span class="o">:::</span><span class="nf">.getTermsDefinition</span><span class="p">(</span><span class="n">terms</span><span class="p">,</span> <span class="nf">ontology</span><span class="p">(</span><span class="n">tg_data</span><span class="p">),</span> <span class="m">65535L</span><span class="p">))</span> <span class="o">%&gt;%</span>
  <span class="n">dplyr</span><span class="o">::</span><span class="nf">relocate</span><span class="p">(</span><span class="n">Term</span><span class="p">,</span> <span class="n">.after</span><span class="o">=</span><span class="n">GO.ID</span><span class="p">)</span> <span class="o">%&gt;%</span>
  <span class="nf">print</span><span class="p">()</span>

<span class="n">tg_table</span> <span class="o">=</span> <span class="n">annoStat</span> <span class="o">%&gt;%</span> <span class="n">dplyr</span><span class="o">::</span><span class="nf">mutate</span><span class="p">(</span>
    <span class="n">classicFisher</span><span class="o">=</span><span class="nf">score</span><span class="p">(</span><span class="n">resClassicFisher</span><span class="p">,</span> <span class="n">whichGO</span><span class="o">=</span><span class="n">GO.ID</span><span class="p">),</span>
    <span class="n">elimFisher</span><span class="o">=</span><span class="nf">score</span><span class="p">(</span><span class="n">resElimFisher</span><span class="p">,</span> <span class="n">whichGO</span><span class="o">=</span><span class="n">GO.ID</span><span class="p">),</span>
    <span class="n">weightFisher</span><span class="o">=</span><span class="nf">score</span><span class="p">(</span><span class="n">resWeightFisher</span><span class="p">,</span> <span class="n">whichGO</span><span class="o">=</span><span class="n">GO.ID</span><span class="p">),</span>
    <span class="n">classicKS</span><span class="o">=</span><span class="nf">score</span><span class="p">(</span><span class="n">resClassicKS</span><span class="p">,</span> <span class="n">whichGO</span><span class="o">=</span><span class="n">GO.ID</span><span class="p">),</span>
    <span class="n">elimKS</span><span class="o">=</span><span class="nf">score</span><span class="p">(</span><span class="n">resElimKS</span><span class="p">,</span> <span class="n">whichGO</span><span class="o">=</span><span class="n">GO.ID</span><span class="p">),</span>
    <span class="n">weightKS</span><span class="o">=</span><span class="nf">score</span><span class="p">(</span><span class="n">resWeightKS</span><span class="p">,</span> <span class="n">whichGO</span><span class="o">=</span><span class="n">GO.ID</span><span class="p">),</span>
  <span class="p">)</span> <span class="o">%&gt;%</span>
  <span class="nf">print</span><span class="p">()</span>
</code></pre></div><p>あとは dplyr, tidyr, ggplot2 などを使って自由に整形、可視化する。</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="n">tg_table</span> <span class="o">%&gt;%</span> <span class="n">dplyr</span><span class="o">::</span><span class="nf">arrange</span><span class="p">(</span><span class="n">elimKS</span><span class="p">)</span>

<span class="n">tg_table</span> <span class="o">%&gt;%</span>
  <span class="n">dplyr</span><span class="o">::</span><span class="nf">select</span><span class="p">(</span><span class="nf">matches</span><span class="p">(</span><span class="s">&#34;Fisher$|KS$&#34;</span><span class="p">))</span> <span class="o">%&gt;%</span>
  <span class="nf">pairs</span><span class="p">()</span>

<span class="n">p0</span> <span class="o">=</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">tg_table</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">geom_point</span><span class="p">(</span><span class="n">shape</span> <span class="o">=</span> <span class="m">16</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="m">0.5</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">coord_fixed</span><span class="p">()</span> <span class="o">+</span>
  <span class="nf">theme_minimal</span><span class="p">()</span>

<span class="n">p0</span> <span class="o">+</span> <span class="nf">aes</span><span class="p">(</span><span class="n">classicFisher</span><span class="p">,</span> <span class="n">classicKS</span><span class="p">)</span>
<span class="n">p0</span> <span class="o">+</span> <span class="nf">aes</span><span class="p">(</span><span class="n">elimFisher</span><span class="p">,</span> <span class="n">elimKS</span><span class="p">)</span>
<span class="n">p0</span> <span class="o">+</span> <span class="nf">aes</span><span class="p">(</span><span class="n">classicKS</span><span class="p">,</span> <span class="n">elimKS</span><span class="p">)</span>
<span class="n">p0</span> <span class="o">+</span> <span class="nf">aes</span><span class="p">(</span><span class="n">classicKS</span><span class="p">,</span> <span class="n">weightKS</span><span class="p">)</span>
<span class="n">p0</span> <span class="o">+</span> <span class="nf">aes</span><span class="p">(</span><span class="n">elimKS</span><span class="p">,</span> <span class="n">weightKS</span><span class="p">)</span>
</code></pre></div><hr>
<p><code>Rgraphviz</code> を使ってDAG描画</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="c1"># BiocManager::install(&#34;Rgraphviz&#34;)</span>
<span class="nf">showSigOfNodes</span><span class="p">(</span><span class="n">tg_data</span><span class="p">,</span> <span class="nf">score</span><span class="p">(</span><span class="n">resElimKS</span><span class="p">),</span> <span class="n">firstSigNodes</span><span class="o">=</span><span class="m">6</span><span class="p">,</span> <span class="n">useInfo</span><span class="o">=</span><span class="s">&#34;all&#34;</span><span class="p">)</span>
<span class="nf">printGraph</span><span class="p">(</span><span class="n">tg_data</span><span class="p">,</span> <span class="n">resElimKS</span><span class="p">,</span> <span class="n">firstSigNodes</span><span class="o">=</span><span class="m">6</span><span class="p">,</span> <span class="n">fn.prefix</span><span class="o">=</span><span class="s">&#34;filename&#34;</span><span class="p">,</span> <span class="n">pdfSW</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div><p>前者はプロットだけ、後者はPDFに書き出し。
significant nodeが四角で、赤いほど低いP値。</p>
<h2 id="go-terms">GO terms</h2>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="n">BPterms</span> <span class="o">=</span> <span class="nf">ls</span><span class="p">(</span><span class="n">GOBPTerm</span><span class="p">)</span>
<span class="n">MFterms</span> <span class="o">=</span> <span class="nf">ls</span><span class="p">(</span><span class="n">GOMFTerm</span><span class="p">)</span>
<span class="n">CCterms</span> <span class="o">=</span> <span class="nf">ls</span><span class="p">(</span><span class="n">GOCCTerm</span><span class="p">)</span>
</code></pre></div>
</article>
</main>

<nav class="side-bar">
<div class="menu">

<div><a href="/about.html">About</a></div>

<div><a href="/research.html">Research Interests</a></div>

<input type="checkbox" id="menu-rstats" checked>
<label for="menu-rstats" class="active">R stats</label>
<ul>
<li><a href="/rstats/intro.html">R自学自習の基礎知識</a></li>
<li><a href="/rstats/programming.html">RプログラミングTips</a></li>
<li><a href="/rstats/config.html">R環境設定</a></li>
<li><a href="/rstats/ggplot2.html">ggplot2</a></li>
<li><a href="/rstats/dplyr.html">dplyr</a></li>
<li><a href="/rstats/tidyr.html">tidyr</a></li>
<li><a href="/rstats/purrr.html">purrr</a></li>
<li><a href="/rstats/readr.html">readr</a></li>
<li><a href="/rstats/stringr.html">stringr</a></li>
<li><a href="/rstats/foreach.html">foreach/parallel</a></li>
<li><a href="/rstats/devtools.html">devtools</a></li>
<li><a href="/rstats/rcpp.html">Rcpp</a></li>
<li><a href="/rstats/bioconductor.html">Bioconductor</a></li>
<li><a href="/rstats/rtracklayer.html">rtracklayer</a></li>
<li><a href="/rstats/biomart.html">biomaRt</a></li>
<li><a href="/rstats/ggbio.html">ggbio</a></li>
<li><a href="/rstats/edger.html">edgeR</a></li>
<li class="active"><a href="/rstats/topgo.html">topGO</a></li>
<li><a href="/rstats/rgl.html">rgl</a></li>
<li><a href="/rstats/stan.html">Stan</a></li>
<li><a href="/rstats/plyr.html">plyr</a></li>
<li><a href="/rstats/reshape2.html">reshape2</a></li>
</ul>

<input type="checkbox" id="menu-python">
<label for="menu-python">Python</label>
<ul>
<li><a href="/python/install.html">Pythonインストール</a></li>
<li><a href="/python/pip.html">pip</a></li>
<li><a href="/python/ipython.html">IPython</a></li>
<li><a href="/python/biopython.html">BioPython</a></li>
<li><a href="/python/concurrent.html">concurrent.futures</a></li>
<li><a href="/python/copy.html">copy</a></li>
<li><a href="/python/matplotlib.html">matplotlib &#43; seaborn</a></li>
<li><a href="/python/scipy.html">NumPy, SciPy</a></li>
<li><a href="/python/pandas.html">Pandas</a></li>
<li><a href="/python/setuptools.html">setuptools</a></li>
</ul>

<input type="checkbox" id="menu-cxx">
<label for="menu-cxx">C&#43;&#43;</label>
<ul>
<li><a href="/cxx/boost.html">Boost</a></li>
<li><a href="/cxx/getopt.html">C&#43;&#43;コマンドライン引数</a></li>
<li><a href="/cxx/speed.html">C&#43;&#43;高速化</a></li>
<li><a href="/cxx/gcc.html">gcc</a></li>
<li><a href="/cxx/bitwise.html">ビット演算</a></li>
<li><a href="/cxx/random.html">擬似乱数生成器</a></li>
</ul>

<input type="checkbox" id="menu-bio">
<label for="menu-bio">Biology</label>
<ul>
<li><a href="/bio/blast.html">BLAST</a></li>
<li><a href="/bio/dadi.html">dadi</a></li>
<li><a href="/bio/gene_ontology.html">Gene Ontology</a></li>
<li><a href="/bio/meme.html">MEME</a></li>
<li><a href="/bio/popgen.html">Population Genetics</a></li>
<li><a href="/bio/repeatmasker.html">RepeatMasker</a></li>
<li><a href="/bio/samtools.html">SAMtools</a></li>
<li><a href="/bio/shirokane.html">SHIROKANE</a></li>
<li><a href="/bio/stochastic_process.html">Stochastic Process</a></li>
<li><a href="/bio/mathmorph.html">数理形態学</a></li>
<li><a href="/bio/linear_algebra.html">線形代数</a></li>
<li><a href="/bio/complexnetwork.html">複雑ネットワーク</a></li>
<li><a href="/bio/nig.html">遺伝研スパコン</a></li>
<li><a href="/bio/motif.html">配列モチーフ探索</a></li>
</ul>

<input type="checkbox" id="menu-dev">
<label for="menu-dev">Developer Tools</label>
<ul>
<li><a href="/dev/atom.html">Atom</a></li>
<li><a href="/dev/autotools.html">autoconf, automake</a></li>
<li><a href="/dev/cmake.html">CMake</a></li>
<li><a href="/dev/emacs.html">Emacs</a></li>
<li><a href="/dev/git.html">Git</a></li>
<li><a href="/dev/make.html">make</a></li>
<li><a href="/dev/mercurial.html">Mercurial</a></li>
<li><a href="/dev/mount.html">mount</a></li>
<li><a href="/dev/nano.html">nano</a></li>
<li><a href="/dev/qt.html">Qt</a></li>
<li><a href="/dev/rsync.html">rsync</a></li>
<li><a href="/dev/ssh.html">ssh</a></li>
<li><a href="/dev/tmux.html">tmux</a></li>
<li><a href="/dev/torque.html">TORQUE</a></li>
<li><a href="/dev/vi.html">vi</a></li>
<li><a href="/dev/zsh.html">zsh</a></li>
<li><a href="/dev/sh.html">シェルスクリプト</a></li>
<li><a href="/dev/nohup.html">プロセス管理</a></li>
<li><a href="/dev/devenv.html">開発環境</a></li>
</ul>

<input type="checkbox" id="menu-mac">
<label for="menu-mac">Mac</label>
<ul>
<li><a href="/mac/applescript.html">AppleScript</a></li>
<li><a href="/mac/homebrew.html">Homebrew</a></li>
<li><a href="/mac/keyboard.html">Keyboard</a></li>
<li><a href="/mac/command.html">Mac固有コマンド</a></li>
<li><a href="/mac/quicklook.html">QuickLook</a></li>
<li><a href="/mac/winebottler.html">WineBottler</a></li>
</ul>

<input type="checkbox" id="menu-lectures">
<label for="menu-lectures">Lectures</label>
<ul>
<li><a href="/lectures/verhoef_comm_ecol-11.html">Community Ecology 輪読会 11章</a></li>
<li><a href="/lectures/git2018nrifs.html">Git入門2018</a></li>
<li><a href="/lectures/git2019makino.html">Git入門2019</a></li>
<li><a href="/lectures/prml-11-1.html">PRML輪読会 11章1節</a></li>
<li><a href="/lectures/prml-2-1.html">PRML輪読会 2章前半</a></li>
<li><a href="/lectures/prml-3-4.html">PRML輪読会 3章4節</a></li>
<li><a href="/lectures/wakeley-2-2.html">Wakeley輪読会 2章2節</a></li>
</ul>

<input type="checkbox" id="menu-misc">
<label for="menu-misc">Miscellaneous</label>
<ul>
<li><a href="/misc/fonts.html">Fonts</a></li>
<li><a href="/misc/gollum.html">Gollum</a></li>
<li><a href="/misc/hugo.html">Hugo</a></li>
<li><a href="/misc/latex.html">LaTeX</a></li>
</ul>

<div><a href="/tags.html">Tags</a></div>
</div>
<form class="cse-search-box" action="/search.html">
<label><input type="text" name="q" aria-label="cse-search-box"></label>
</form>
</nav>
<footer><small>(ɔ) 2008 岩嵜航 Watal M. Iwasaki</small></footer>
</body>
</html>
