<!DOCTYPE html>
<html lang="ja">
<head>
<meta charset="utf-8">
<meta name="generator" content="Hugo 0.16-DEV" />
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,400italic,700,700italic%7CUbuntu+Mono:400,400italic,700,700italic">
<link rel="stylesheet" href="/css/theme.css">
<link rel="stylesheet" href="/css/highlight.css">
<link rel="shortcut icon" href="/favicon.ico">
<link rel="icon" href="/favicon-192x192.png" sizes="192x192">
<link rel="apple-touch-icon" href="/favicon-192x192.png" sizes="192x192">
<title>BioPython: Tools for biological computation - Watal M. Iwasaki</title>
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    inlineMath: [['$','$']],
    displayMath: [['$$','$$']],
    processEnvironments: false,
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
MathJax.Hub.Queue(function() {
  var all = MathJax.Hub.getAllJax(), i;
  for(i = 0; i < all.length; i += 1) {
    all[i].SourceElement().parentNode.classList.add('has-jax');
  }
});
</script>
<script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_CHTML"></script>

<script>
(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
})(window,document,'script','//www.google-analytics.com/analytics.js','ga');
ga('create', 'UA-41178626-2', 'auto');
ga('send', 'pageview');
</script>
</head>
<body><div id="container">
<header><h1><a href="/">
<img class="logo" src="/favicon-192x192.png" alt="航">
Watal M. Iwasaki
</a></h1>
</header>

<main>
<article>
<header><h1><a href="">
BioPython — Tools for biological computation
</a></h1>
<ul id="tags">
<li><a href="/tags/python">python</a></li>
</ul>
</header>



<p><a href="http://biopython.org">http://biopython.org</a>
<a href="http://biopython.org/DIST/docs/api/Bio-module.html">http://biopython.org/DIST/docs/api/Bio-module.html</a></p>

<h2 id="seqio"><code>SeqIO</code></h2>

<ul>
<li><a href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a></li>
<li><a href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec51">http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec51</a></li>
<li><a href="http://biopython.org/DIST/docs/api/Bio.SeqIO-module.html">http://biopython.org/DIST/docs/api/Bio.SeqIO-module.html</a></li>
</ul>

<dl>
<dt><code>read(file, format, alphabet=None)</code></dt>
<dd>1配列しか含まないファイルを読んで <code>SeqRecord</code> を返す</dd>
<dt><code>parse(file, format, alphabet=None)</code></dt>
<dd>複数の配列を含むファイルを読んで <code>SeqRecord</code> のイテレータを返す</dd>
<dt><code>index(filename, format, alphabet=None, key_function=None)</code></dt>
<dd><code>id</code> をキーとした辞書likeオブジェクトを返す</dd>
</dl>

<p><code>write(sequences, file, format)</code></p>

<p><code>to_dict(sequences, key_function=None)</code></p>

<p><code>convert(in_file, in_format, out_file, out_format, alphabet=None)</code></p>

<pre><code>from Bio import SeqIO

with open('beer.fasta', 'r') as fin:
    for record in SeqIO.parse(fin, 'fasta'):
        print(record.id)
        print(record.seq)
</code></pre>

<h2 id="seqrecord"><code>SeqRecord</code></h2>

<ul>
<li><a href="http://biopython.org/wiki/SeqRecord">http://biopython.org/wiki/SeqRecord</a></li>
<li><a href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec32">http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec32</a></li>
<li><a href="http://biopython.org/DIST/docs/api/Bio.SeqRecord.SeqRecord-class.html">http://biopython.org/DIST/docs/api/Bio.SeqRecord.SeqRecord-class.html</a></li>
</ul>

<p><code>Seq</code> とそのほかの情報をひとまとまりにしたクラス。</p>

<p><code>id</code><br />
<code>seq</code></p>

<p><code>name</code><br />
<code>description</code><br />
<code>dbxrefs</code><br />
<code>features</code><br />
<code>annotations</code><br />
<code>letter_annotations</code></p>

<pre><code>&gt;&gt;&gt; for record in SeqIO.parse(fin, 'fasta'):
...     print(record)
...
ID: gi|186972394|gb|EU490707.1|
Name: gi|186972394|gb|EU490707.1|
Description: gi|186972394|gb|EU490707.1| Selenipedium aequinoctiale maturase K (matK) gene, partial cds; chloroplast
Number of features: 0
Seq('ATTTTTTACGAACCTGTGGAAATTTTTGGTTATGACAATAAATCTAGTTTAGTA...GAA', SingleLetterAlphabet())
ID: gi|186972391|gb|ACC99454.1|
Name: gi|186972391|gb|ACC99454.1|
Description: gi|186972391|gb|ACC99454.1| maturase K [Scaphosepalum rapax]
Number of features: 0
Seq('IFYEPVEILGYDNKSSLVLVKRLITRMYQQKSLISSLNDSNQNEFWGHKNSFSS...EEE', SingleLetterAlphabet())
</code></pre>

<h2 id="seq"><code>Seq</code></h2>

<ul>
<li><a href="http://biopython.org/wiki/Seq">http://biopython.org/wiki/Seq</a></li>
<li><a href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec17">http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec17</a></li>
<li><a href="http://biopython.org/DIST/docs/api/Bio.Seq.Seq-class.html">http://biopython.org/DIST/docs/api/Bio.Seq.Seq-class.html</a></li>
</ul>

<p>塩基配列・アミノ酸配列のクラス。
標準 <code>str</code> とほとんど同じように扱えるほか、
相補鎖とかなんとかを簡単に扱えるメソッドが備わっている。</p>

<ul>
<li><code>complement(self)</code></li>
<li><code>reverse_complement(self)</code></li>
<li><code>transcribe(self)</code></li>
<li><code>back_transcribe(self)</code></li>
<li><code>translate(self, table=&quot;Standard&quot;, stop_symbol=&quot;*&quot;, to_stop=False, cds=False)</code></li>
<li><code>ungap(self, gap=None)</code></li>
</ul>

<h2 id="genomediagram">GenomeDiagram</h2>

<p><a href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec328">http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec328</a>
<a href="http://biopython.org/DIST/docs/GenomeDiagram/userguide.pdf">http://biopython.org/DIST/docs/GenomeDiagram/userguide.pdf</a></p>

<h2 id="entrez">Entrez</h2>

<ul>
<li><a href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc90">http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc90</a></li>
</ul>

<h3 id="esearch">esearch</h3>

<ul>
<li><a href="http://www.ncbi.nlm.nih.gov/entrez/query/static/esearch_help.html">NCBI ESerch Utility</a></li>
<li><a href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc93">http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc93</a></li>
</ul>

<!-- -->

<pre><code>&gt;&gt;&gt; from Bio import Entrez
&gt;&gt;&gt; handle = Entrez.esearch(db=&quot;nucleotide&quot;, retmax=10, term=&quot;Opuntia&quot;)
&gt;&gt;&gt; record = Entrez.read(handle)
&gt;&gt;&gt; for item in record.items():
...     print(item)
...
(u'Count', '390')
(u'RetMax', '10')
(u'IdList', ['257359511', '283467266', '246905625', '246905624', '246655205', '246655204', '240253899', '240253897', '240253576', '240253574'])
(u'TranslationStack', [{u'Count': '200', u'Field': 'Organism', u'Term': '&quot;Opuntia&quot;[Organism]', u'Explode': 'Y'}, {u'Count': '390', u'Field': 'All Fields', u'Term': 'Opuntia[All Fields]', u'Explode': 'Y'}, 'OR', 'GROUP'])
(u'TranslationSet', [{u'To': '&quot;Opuntia&quot;[Organism] OR Opuntia[All Fields]', u'From': 'Opuntia'}])
(u'RetStart', '0')
(u'QueryTranslation', '&quot;Opuntia&quot;[Organism] OR Opuntia[All Fields]')
</code></pre>

<h3 id="efetch">efetch</h3>

<ul>
<li><a href="http://eutils.ncbi.nlm.nih.gov/corehtml/query/static/efetch_help.html">NCBI EFetch Utility</a></li>
<li><a href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc96">http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc96</a></li>
</ul>

<p>返り値は結果(XML)へのハンドル。:</p>

<pre><code>from Bio import Entrez

handle = Entrez.efetch(db=&quot;nucleotide&quot;, id=&quot;186972394,186972394&quot;, rettype=&quot;fasta&quot;)
record = SeqIO.parse(handle, &quot;fasta&quot;)
for x in record:
    print(i)
</code></pre>

<h2 id="installation">Installation</h2>

<p><a href="/python/pip.html">pip</a> で一発:</p>

<pre><code>% pip install biopython
</code></pre>

</article>
</main>

<nav id="menu">

<div><a href="/cv.html">Curriculum Vitae</a></div>


<div><a href="/research.html">Research Interests</a></div>


<input type="checkbox" id="rstats">
<label for="rstats">R stats</label>
<ul>
<li><a href="/rstats/bioconductor.html">Bioconductor</a></li>
<li><a href="/rstats/biomart.html">biomaRt</a></li>
<li><a href="/rstats/devtools.html">devtools</a></li>
<li><a href="/rstats/dplyr.html">dplyr</a></li>
<li><a href="/rstats/edger.html">edgeR</a></li>
<li><a href="/rstats/ggbio.html">ggbio</a></li>
<li><a href="/rstats/ggplot2.html">ggplot2</a></li>
<li><a href="/rstats/plyr.html">plyr</a></li>
<li><a href="/rstats/readr.html">readr</a></li>
<li><a href="/rstats/reshape2.html">reshape2</a></li>
<li><a href="/rstats/rgl.html">rgl</a></li>
<li><a href="/rstats/rjags.html">rjags</a></li>
<li><a href="/rstats/rtracklayer.html">rtracklayer</a></li>
<li><a href="/rstats/stringr.html">stringr</a></li>
<li><a href="/rstats/tidyr.html">tidyr</a></li>
<li><a href="/rstats/topgo.html">topGO</a></li>
<li><a href="/rstats/programming.html">プログラミングTips</a></li>
<li><a href="/rstats/config.html">環境設定</a></li>
<li><a href="/rstats/intro.html">自学自習のための基礎知識</a></li>
</ul>

<input type="checkbox" id="python" checked>
<label for="python" class="active">Python</label>
<ul>
<li class="active"><a href="/python/biopython.html">BioPython</a></li>
<li><a href="/python/egglib.html">EggLib</a></li>
<li><a href="/python/ipython.html">IPython</a></li>
<li><a href="/python/install.html">Installation</a></li>
<li><a href="/python/scipy.html">NumPy, SciPy</a></li>
<li><a href="/python/pyqt.html">PyQt</a></li>
<li><a href="/python/copy.html">copy</a></li>
<li><a href="/python/matplotlib.html">matplotlib &#43; seaborn</a></li>
<li><a href="/python/pip.html">pip</a></li>
<li><a href="/python/rpy2.html">rpy2</a></li>
</ul>

<input type="checkbox" id="cxx">
<label for="cxx">C&#43;&#43;</label>
<ul>
<li><a href="/cxx/boost.html">Boost</a></li>
<li><a href="/cxx/getopt.html">C&#43;&#43; コマンドライン引数の取得</a></li>
<li><a href="/cxx/clang.html">clang / llvm</a></li>
<li><a href="/cxx/gcc.html">gcc</a></li>
<li><a href="/cxx/speed.html">めざせC&#43;&#43;高速プログラム</a></li>
<li><a href="/cxx/random.html">擬似乱数生成器</a></li>
</ul>

<input type="checkbox" id="bio">
<label for="bio">Biology</label>
<ul>
<li><a href="/bio/blast.html">BLAST</a></li>
<li><a href="/bio/dnasp.html">DnaSP</a></li>
<li><a href="/bio/emboss.html">EMBOSS</a></li>
<li><a href="/bio/ensembl.html">Ensembl</a></li>
<li><a href="/bio/gene_ontology.html">Gene Ontology</a></li>
<li><a href="/bio/meme.html">MEME</a></li>
<li><a href="/bio/mrbayes.html">MrBayes</a></li>
<li><a href="/bio/paml.html">PAML</a></li>
<li><a href="/bio/popgen.html">Population Genetics</a></li>
<li><a href="/bio/repeatmasker.html">RepeatMasker</a></li>
<li><a href="/bio/samtools.html">SAMtools</a></li>
<li><a href="/bio/stochastic_process.html">Stochastic Process</a></li>
<li><a href="/bio/dadi.html">dadi</a></li>
<li><a href="/bio/linear_algebra.html">線形代数</a></li>
<li><a href="/bio/complexnetwork.html">複雑ネットワーク</a></li>
<li><a href="/bio/nig.html">遺伝研スパコン</a></li>
<li><a href="/bio/motif.html">配列モチーフ探索</a></li>
</ul>

<input type="checkbox" id="dev">
<label for="dev">Developer Tools</label>
<ul>
<li><a href="/dev/etc.html">/etc</a></li>
<li><a href="/dev/emacs.html">Emacs</a></li>
<li><a href="/dev/mercurial.html">Mercurial</a></li>
<li><a href="/dev/mysql.html">MySQL</a></li>
<li><a href="/dev/qt.html">Qt</a></li>
<li><a href="/dev/sh.html">Shell Script</a></li>
<li><a href="/dev/torque.html">TORQUE</a></li>
<li><a href="/dev/tex.html">TeX</a></li>
<li><a href="/dev/autotools.html">autoconf, automake</a></li>
<li><a href="/dev/make.html">make</a></li>
<li><a href="/dev/mount.html">mount</a></li>
<li><a href="/dev/nano.html">nano</a></li>
<li><a href="/dev/rsync.html">rsync</a></li>
<li><a href="/dev/ssh.html">ssh</a></li>
<li><a href="/dev/sshfs.html">sshfs</a></li>
<li><a href="/dev/tmux.html">tmux</a></li>
<li><a href="/dev/zsh.html">zsh</a></li>
<li><a href="/dev/nohup.html">プロセス管理</a></li>
<li><a href="/dev/devenv.html">開発環境</a></li>
</ul>

<input type="checkbox" id="linux">
<label for="linux">Linux</label>
<ul>
<li><a href="/linux/centos.html">CentOS 6.5</a></li>
<li><a href="/linux/japanese.html">Linux日本語環境</a></li>
<li><a href="/linux/apt.html">apt/dpkg</a></li>
<li><a href="/linux/ufw.html">ufw</a></li>
</ul>

<input type="checkbox" id="mac">
<label for="mac">Mac</label>
<ul>
<li><a href="/mac/applescript.html">AppleScript</a></li>
<li><a href="/mac/homebrew.html">Homebrew</a></li>
<li><a href="/mac/keyboard.html">Keyboard</a></li>
<li><a href="/mac/command.html">Mac Command</a></li>
<li><a href="/mac/macports.html">MacPorts</a></li>
<li><a href="/mac/quicklook.html">QuickLook</a></li>
<li><a href="/mac/spotlight.html">Spotlight</a></li>
<li><a href="/mac/winebottler.html">WineBottler</a></li>
<li><a href="/mac/kotoeri.html">ことえり</a></li>
</ul>

<input type="checkbox" id="lectures">
<label for="lectures">Lectures</label>
<ul>
<li><a href="/lectures/verhoef_comm_ecol-11.html">Community Ecology 輪読会 11章</a></li>
<li><a href="/lectures/prml-11-1.html">PRML輪読会 11章1節</a></li>
<li><a href="/lectures/prml-2-1.html">PRML輪読会 2章前半</a></li>
<li><a href="/lectures/prml-3-4.html">PRML輪読会 3章4節</a></li>
<li><a href="/lectures/wakeley-2-2.html">Wakeley輪読会 2章2節</a></li>
</ul>

<input type="checkbox" id="misc">
<label for="misc">Miscellaneous</label>
<ul>
<li><a href="/misc/fonts.html">Fonts</a></li>
<li><a href="/misc/mailman.html">Mailman</a></li>
<li><a href="/misc/vnc.html">VNCによる画面共有</a></li>
<li><a href="/misc/virtualbox.html">VirtualBox</a></li>
</ul>

<div><a href="/tags">Tags</a></div>

</nav>




<footer>(ɔ) 2008 Watal M. Iwasaki</footer>
</div>

<script src="/js/highlight.pack.js"></script>
<script>
hljs.configure({languages: ["sh","c++","python","r","tex"]});
hljs.initHighlightingOnLoad();
</script>
</body>
</html>

